### PR TITLE
AES-XTS Enc Dec test on rand incremental length inputs

### PR DESCRIPTION
**Pull request by manastasova**
Add incremental length random input message tests for AES-XTS encryption/decryption

## Issues:
N/A

## Description of changes:
AWS-LC's current AES-XTS tests use fixed-size test vectors but do not systematically verify that encryption and decryption work correctly across different message lengths.

This PR adds a new test that generates random inputs of incremental lengths and verify that `encrypt(decrypt(input)) == input` for each length. This ensures that AES-XTS correctly handles messages of varying sizes, particularly:
- Messages at and above the minimum length (16 bytes)
- Messages that require ciphertext stealing (non-block-aligned sizes)
- Messages of various sizes up to the practical maximum

The tests also add explicit length equality checks between plaintext and ciphertext to catch potential length-related bugs early.

## Call-outs:
The test provides better confidence in correctness across different input sizes.

## Testing:
`./crypto/crypto_test --gtest_filter=XTSTest.EncryptDecryptRand.`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: crypto/fipsmodule/modes/xts_test.cc [hunk 1]
```diff
@@ -1240,6 +1244,90 @@ TEST(XTSTest, TestVectors) {
 #endif
 }
 
+TEST(XTSTest, EncryptDecyptRand) {
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: Instead of being specific to XTS. Can expand to cover all the cipher implementations similar to tests in `cipher_test.cc` https://github.com/aws/aws-lc/blob/main/crypto/cipher_extra/cipher_test.cc#L543-L589

They use known input-output test files though. So, would need to extend that to randomised testing.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 2]
```diff
@@ -1240,6 +1244,90 @@ TEST(XTSTest, TestVectors) {
...
+        if (in_place) {
+          out_p = in_p;
+        }
+
+        bssl::ScopedEVP_CIPHER_CTX ctx;
+        ASSERT_TRUE(EVP_EncryptInit_ex(ctx.get(), cipher, nullptr, key.data(),
+                                      iv.data()));
+        ASSERT_TRUE(
+            EVP_EncryptUpdate(ctx.get(), out_p, &len, in_p, plaintext.size()));
+        OPENSSL_memcpy(ciphertext.data(), out_p, plaintext.size());
```
Comment (nebeid) [inline crypto/fipsmodule/modes/xts_test.cc]: Is this line, and more so `ciphertext`, used?

File: crypto/fipsmodule/modes/xts_test.cc (hunk 1 above)
Comment (nebeid) [inline crypto/fipsmodule/modes/xts_test.cc]:
```suggestion
TEST(XTSTest, EncryptDecryptRand) {
```

File: crypto/fipsmodule/modes/xts_test.cc (hunk 2 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: Indeed, not used. Thanks, Nevine. Updated.

File: crypto/fipsmodule/modes/xts_test.cc (hunk 1 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: Thanks!

File: crypto/fipsmodule/modes/xts_test.cc [hunk 3]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+#if defined(OPENSSL_LINUX)
+  int pagesize = sysconf(_SC_PAGE_SIZE);
+  ASSERT_GE(pagesize, 0);
+  uint8_t *in_buffer_beg = get_buffer_beg(pagesize);
+  uint8_t *out_buffer_beg = get_buffer_beg(pagesize);
+  uint8_t *in_buffer_end = in_buffer_beg + pagesize;
+  uint8_t *out_buffer_end = out_buffer_beg + pagesize;
+#endif
+
+const EVP_CIPHER *cipher = EVP_aes_256_xts();
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: indentation

File: crypto/fipsmodule/modes/xts_test.cc [hunk 4]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+      if (!beg) {
+        in_p = in_buffer_end - plaintext.size();
+        out_p = out_buffer_end - plaintext.size();
+      } else {
+        in_p = in_buffer_end - pagesize;
+        out_p = out_buffer_end - pagesize;
+      }
+      OPENSSL_memset(in_p, 0x00, plaintext.size());
+      OPENSSL_memset(out_p, 0x00, plaintext.size());
+    #else
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: indentation

File: crypto/fipsmodule/modes/xts_test.cc [hunk 5]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+        }
+
+        ctx.Reset();
+        len = 0;
+        ASSERT_TRUE(EVP_DecryptInit_ex(ctx.get(), cipher, nullptr, key.data(),
+                                      iv.data()));
+        ASSERT_TRUE(
+            EVP_DecryptUpdate(ctx.get(), in_p, &len, out_p, plaintext.size()));
+
+        decrypted_len = len;
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: nitpicking, but variables `decrypted_len` and `encrypted_len` seems redundant

File: crypto/fipsmodule/modes/xts_test.cc [hunk 6]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+        ASSERT_EQ(decrypted_len, plaintext.size());
+        EXPECT_EQ(Bytes(plaintext), Bytes(in_p, static_cast<size_t>(len)));
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc:1336]: might as well also be an assert.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 7]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+  uint8_t *in_buffer_end = in_buffer_beg + pagesize;
+  uint8_t *out_buffer_end = out_buffer_beg + pagesize;
+#endif
+
+const EVP_CIPHER *cipher = EVP_aes_256_xts();
+
+  // Test AESXTS Encrypt and Decrypt with random messages of incremental lenghts
+  for (size_t msg_len = 16; msg_len < AESXTS_RAND_MSG_MAX_LEN ; msg_len += 1) {
+
+    std::vector<uint8_t> key(EVP_CIPHER_key_length(cipher)), iv(EVP_CIPHER_iv_length(cipher)),
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: nitpicking, but don't need to allocate new memory per loop iteration. (same for `in` and `out` below). In the grand scheme of things, won't save much testing time though.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 8]
```diff
@@ -23,6 +25,8 @@
 #include "internal.h"
 #include "../../test/test_util.h"
 
+#define AESXTS_RAND_MSG_MAX_LEN 4096
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]:
```suggestion
static constexpr size_t AESXTS_RAND_MSG_MAX_LEN = 4096
```
for type-safety and more c++ idiomatic

File: crypto/fipsmodule/modes/xts_test.cc [hunk 9]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+                                      iv.data()));
+        ASSERT_TRUE(
+            EVP_EncryptUpdate(ctx.get(), out_p, &len, in_p, plaintext.size()));
+
+        encrypted_len = len;
+        ASSERT_EQ(encrypted_len, plaintext.size());
+
+        // Test decryption.
+
+        if (!in_place) {
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: if-condition not needed.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 10]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+      in_p = in.get();
+      out_p = out.get();
+    #endif
+      for (bool in_place : {false, true}) {
+        SCOPED_TRACE(in_place);
+
+        // Test encryption.
+
+        OPENSSL_memcpy(in_p, plaintext.data(), plaintext.size());
+        if (in_place) {
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc:1311]: Just a note: The not in-place test case assumes that the loop initializer list being processed in-order from left-to-right. Which it is! If not, we would always test the in-place case.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 11]
```diff
@@ -1240,6 +1244,95 @@ TEST(XTSTest, TestVectors) {
...
+    RAND_bytes(iv.data(), EVP_CIPHER_iv_length(cipher));
+    RAND_bytes(plaintext.data(), msg_len);
+
+    SCOPED_TRACE(plaintext.size());
+
+    int len = 0;
+    size_t decrypted_len = 0, encrypted_len = 0;
+    uint8_t *in_p = nullptr, *out_p = nullptr;
+  #if defined(OPENSSL_LINUX)
+    ASSERT_GE(pagesize, (int)plaintext.size());
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: Hopefully, we will never meet an system that requires small pagesizes (e.g. below 4kib). I understand the test here. But maybe we can just take the minimal of pagesize and max plaintext size and let that be the upper bound we test to? Slightly more robust.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 12]
```diff
@@ -15,6 +15,8 @@
 #include <openssl/cipher.h>
 #include <openssl/aes.h>
```
Comment (torben-hansen) [inline crypto/fipsmodule/modes/xts_test.cc]: No newline

File: crypto/fipsmodule/modes/xts_test.cc (hunk 8 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: Thank you

File: crypto/fipsmodule/modes/xts_test.cc (hunk 10 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc:1311]: I agree

File: crypto/fipsmodule/modes/xts_test.cc (hunk 7 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: `in`, `out` and `plaintext` are of different lengths based on the current loop iteration. I moved `key` and `iv` outside of the loop.

File: crypto/fipsmodule/modes/xts_test.cc (hunk 11 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: I see your point (even though it is, indeed, very unlikely to have linux page < 4kib). 
I still believe it is useful to explicitly test the edge cases related to page size. I keep it as is with an additional check if OPENSSL_LINUX and pagesize < plaintext.size();

File: crypto/fipsmodule/modes/xts_test.cc (hunk 5 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: I agree! Fixed! Thanks!

File: crypto/fipsmodule/modes/xts_test.cc (hunk 6 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc:1336]: I prefer to keep it as `EXPECT_EQ` so that all the failing tests can be listed and tracked (e.g., plaintext values between 17 and 31 bytes will all fail, but will pass if length is above 31 bytes - adds clues for debugging).

File: crypto/fipsmodule/modes/xts_test.cc (hunk 12 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: thanks

File: crypto/fipsmodule/modes/xts_test.cc (hunk 3 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: fixed

File: crypto/fipsmodule/modes/xts_test.cc (hunk 4 above)
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc]: thanks

File: crypto/fipsmodule/modes/xts_test.cc (hunk 6 above)
Comment (torben-hansen) [reply crypto/fipsmodule/modes/xts_test.cc:1336]: Sure - Just noted that all other tests are asserts.

File: crypto/fipsmodule/modes/xts_test.cc (hunk 7 above)
Comment (torben-hansen) [reply crypto/fipsmodule/modes/xts_test.cc]: Right, could be allocated with a length upper bound (:
But can leave as-is.

File: crypto/fipsmodule/modes/xts_test.cc [hunk 13]
```diff
@@ -1240,6 +1243,104 @@ TEST(XTSTest, TestVectors) {
...
+      for (bool in_place : {false, true}) {
+        SCOPED_TRACE(in_place);
+
+        // Generate random key and iv for each encryption test
+        RAND_bytes(key.data(), EVP_CIPHER_key_length(cipher));
+        RAND_bytes(iv.data(), EVP_CIPHER_iv_length(cipher));
+
+        // Test encryption.
+
+        OPENSSL_memcpy(in_p, plaintext.data(), plaintext.size());
```
[the remark this answers is not in the document]
Comment (manastasova) [reply crypto/fipsmodule/modes/xts_test.cc:1310]: Not sure why this warning appears. `in_p` is used as the input message for the encryption function later on.

**Review by nebeid — APPROVED** [review]
